### PR TITLE
OCI support in Helm builtin

### PR DESCRIPTION
**Pull request by mikebz**
This is currently WIP.  

fixes: #4381 

Questions for owner: @natasha41575 @yuwenma @KnVerey 
It seems like @monopole has added environment variables to put the data and cache for HELM into a temp directory.  Do we know why that is?  The reason why this is undesirable with OCI is that most of the OCI repos are protected with auth.  Auth is possible if one does authentication with Helm ahead of time.  Example:

> helm registry login -u _json_key --password-stdin https://us-central1-docker.pkg.dev

After that is run I am able to do things like:
```
helmCharts:
- name: chart1
  version: 0.1.0
  repo: oci://us-central1-docker.pkg.dev/mikebz-ex1/charts
  valuesInline:
    nameOverride: foobar
```
and if the configuration is not in the temp location I can easily get the chart.

One of the ideas is to include some sort of an authentication in kustomization.yaml, but I think that would be undesirable.

Things that are left to figure out:

- [ ] create a publicly available registry or something that can be used to ensure that the unit test works.
- [ ] figure out if unsetting the HELM* variables will have a security or other undesirable impact.
- [ ] is there an issue updating the default Helm download to 3.8+?  OCI is enabled by default in those builds

### CODE REVIEW

**Comment by natasha41575** [comment]
This PR is probably more appropriate to make in this repo: https://github.com/kubernetes-sigs/krm-functions-registry/tree/main/krm-functions/sig-cli/render-helm-chart

On the kustomize side, we have paused development on the helm plugin until we are able to migrate to the KRM function. The long term plan is here: https://github.com/kubernetes-sigs/kustomize/issues/4401 and work is in progress on this front

**Comment by natasha41575** [comment]
/hold

Having the test infra will be great, but it's also been pointed out to me that this is a hotfix for a particular use case. I'm designing a more general solution that we will put into the KRM function, after which we can see if we want to put it in the builtin plugin. But, this is a more substantial feature than I'd realized when I first saw this PR so I believe a better course of action would be to prioritize migration (which I plan to do this quarter).

**Comment by natasha41575** [comment]
An update from the infra folks:

The k8s infra doesn't yet properly support Artifact Registry, and they are hesitant to provision it for a single ad-hoc use case that the kustomize maintainers are skeptical about. It is more realistic for them to support it for us for the KRM functions registry a few months from now.

**Comment by mikebz** [comment]
Thanks for the update.  One of the things to consider is that the current tests are hitting 3rd party chart registries.  Maybe we can create one for functions and still use it for built in and the function.  There is no precedent right now that all of the test repositories are CNCF owned.

File: plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go
Comment (tis-rpage) [inline plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go:302]: Just to clarify, it's saying `Url` should be `URL` in the variable name, I missed the distinction on first read-through.

**Comment by tis-rpage** [comment]
> /hold
> 
> Having the test infra will be great, but it's also been pointed out to me that this is a hotfix for a particular use case. I'm designing a more general solution that we will put into the KRM function, after which we can see if we want to put it in the builtin plugin. But, this is a more substantial feature than I'd realized when I first saw this PR so I believe a better course of action would be to prioritize migration (which I plan to do this quarter).
> 
> Again, the helm builtin is slated for deprecation and eventual removal, so our support on it is extremely limited.

Is it truly unacceptable to accept this patch as a stop-gap measure to unblock the user base of kustomize from using OCI registries?

I agree the KRM function will be better long-term, but as it stands, the user base has to run external commands to pull OCI charts and make them available to kustomize locally before its invocation. Seeing as the helm team broke semver compliance, it stands to reason a hotfix should be acceptable to address the upstream products bad behavior.

**Comment by bgrant0607** [comment]
KRM functions aren't yet a viable replacement for built-in plugins. For instance, of the GitOps tools that support kustomize, which ones support KRM functions? Are there examples of executing KRM functions in CI?

There was a suggestion to fork kustomize here:
https://github.com/argoproj/argo-cd/issues/5553#issuecomment-1028529945

Since there are uses of the helm built-in plugin, I recommend continuing to support it until KRM function execution obstacles are overcome.

**Comment by bgrant0607** [comment]
Example user: https://twitter.com/todaywasawesome/status/1532078601448787968

**Comment by igstbagusdharmaputra** [comment]
Waiting for closed this issue 😁

**Comment by mikebz** [comment]
I am guessing this addition is unwelcome.  We have already created a workaround in ConfigSync, so my interest in lobbying for this more has waned :)

**Comment by igstbagusdharmaputra** [comment]
> I am guessing this addition is unwelcome. We have already created a workaround in ConfigSync, so my interest in lobbying for this more has waned :)

i hope this issue can solving, because i need pull chart from oci repository

**Comment by igstbagusdharmaputra** [comment]
I hope this issue can solving, because i need pull oci registry from ArgoCD

Pada tanggal Jum, 14 Okt 2022 23.51, Mike Borozdin ***@***.***>
menulis:

**Comment by antonydevanchi** [comment]
@mikebz bump?

**Review by diegoluisi — APPROVED** [review]

**Comment by antonydevanchi** [comment]
@diegoluisi @mikebz thank you for yours job done earlier. What next steps? It's not clear from last activity in thread.

**Comment by mikebz** [comment]
I think we need to clear go ahead from the project maintainers @KnVerey
There is a concern here about backwards compatibility but for an undocumented feature.

**Comment by antonydevanchi** [comment]
@mikebz can I help something?

Also I did PR with fix lint alert: https://github.com/mikebz/kustomize/pull/1/files
If it's useles, so then just feel free just drop request :)

**Comment by mikebz** [comment]
Thank you @antonydevanchi !  There are a few things to do to get this PR back into shape.  I think the major question is whether we introduce a potential breaking change or not.

**Comment by natasha41575** [comment]
@KnVerey and I discussed this, and the primary concern we have is a lack of bandwidth. There are a few things we need to address here, one of which is auth.

> It seems like @monopole has added environment variables to put the data and cache for HELM into a temp directory. Do we know why that is? 

I believe this is because of kustomize's foundational principle that the kustomization file should not depend on its environment, which includes that it should not be able to get auth information from its environment. Breaking this encapsulation by allowing the user to run `helm registry login -u _json_key --password-stdin https://us-central1-docker.pkg.dev/` prior to `kustomize build` may work but it would add the kustomize eschewed feature of [build time side effects](https://kubectl.docs.kubernetes.io/faq/kustomize/eschewedfeatures/#build-time-side-effects-from-cli-args-or-env-variables).

Because of that, we cannot accept the PR as is. OCI support in the Helm builtin then requires more thoughtful design, in particular how to properly do auth. As such, it would require a[ kustomize mini-KEP](https://github.com/kubernetes-sigs/kustomize/tree/master/proposals).

For transparency, there are only 3-4 people working on kustomize, most of whom are spending less than 10% of their time on it. We need to prioritize things that are core to kustomize functionality, and additionally, none of us are particularly familiar with helm. We already have a large number of open PRs and one open KEP that we are struggling to find time to move forward, and as a result, we cannot promise that we have the bandwidth to support this feature.

**Comment by tis-rpage** [comment]
> I believe this is because of kustomize's foundational principle that the kustomization file should not depend on its environment, which includes that it should not be able to get auth information from its environment. Breaking this encapsulation by allowing the user to run `helm registry login -u _json_key --password-stdin https://us-central1-docker.pkg.dev/` prior to `kustomize build` may work but it would add the kustomize eschewed feature of [build time side effects](https://kubectl.docs.kubernetes.io/faq/kustomize/eschewedfeatures/#build-time-side-effects-from-cli-args-or-env-variables).

If this is truly the case, kustomize needs to break the ability to support network references that are fetched at build time, making the tool far less valuable when external network resources for kustomizations and helm charts can no longer be referenced, which by necessity are build time side-effects of a network fetch at a point in time.

**Pull request closed without merging**
